src/security/ldap/src/main/java/org/geoserver/security/ldap/BindingLdapAuthoritiesPopulator.java

```diff
@@ -186,33 +216,123 @@ public void executeWithContext(
+            String groupDn,
+            String groupName,
+            Set<GrantedAuthority> authorities,
+            int depth) {
+
+        if (logger.isDebugEnabled()) {
+            logger.debug(
+                    "Searching for roles for nested group '"
+                            + groupName
+                            + "', DN = "
+                            + "'"
+                            + groupDn
+                            + "', with filter "
+                            + nestedGroupSearchFilter
+                            + " in search base '"
+                            + getGroupSearchBase()
+                            + "'");
+        }
+
+        final List<Pair<String, String>> groupRolesNameDn = new ArrayList<>();
+        ctxConsumer.accept(
+                (ctx) -> {
+                    SpringSecurityLdapTemplate authTemplate =
+                            (SpringSecurityLdapTemplate)
+                                    LDAPUtils.getLdapTemplateInContext(ctx, ldapTemplate);
+                    // Get ldap groups in form of Pair<String,String> -> Pair<name,dn>
+                    final String formattedFilter =
+                            MessageFormat.format(nestedGroupSearchFilter, groupDn, groupName);
+                    groupRolesNameDn.addAll(
+                            authTemplate.search(
+                                    getGroupSearchBase(),
+                                    formattedFilter,
+                                    new AbstractContextMapper<Pair<String, String>>() {
+                                        @Override
+                                        protected Pair<String, String> doMapFromContext(
+                                                DirContextOperations ctx) {
+                                            String name =
+                                                    ctx.getStringAttribute(groupRoleAttribute);
+                                            String dn = ctx.getNameInNamespace();
+                                            return Pair.of(name, dn);
+                                        }
+                                    }));
+                });
+
+        if (logger.isDebugEnabled()) {
+            logger.debug("Roles from search: " + groupRolesNameDn);
+        }
+
+        for (Pair<String, String> roleNameDn : groupRolesNameDn) {
+            String role = roleNameDn.getLeft();
+            String dn = roleNameDn.getRight();
+            if (convertToUpperCase) {
+                role = role.toUpperCase();
+            }
+            boolean addedSuccesfuly =
+                    authorities.add(new SimpleGrantedAuthority(rolePrefix + role));
+            // search nested Ldap groups,
+            // only if role was added successfully for avoiding circular references
+            // if maxGroupSearchLevel == -1 -> no depth limit
+            if (maxGroupSearchLevel == -1 || (depth > 0 && addedSuccesfuly))
```

Comment:
So if there is no depth limit, we don't care about circular dependencies?

Reply:
Right, fixed. Thanks.